pkg/controller/volume/attachdetach/attach_detach_controller.go

```diff
@@ -649,6 +667,65 @@ func (adc *attachDetachController) processVolumesInUse(
 	}
 }
 
+// installCRDs creates the specified CustomResourceDefinition for the CSIDrivers object.
+func (adc *attachDetachController) installCRDs() error {
+	crd := &apiextensionsv1beta1.CustomResourceDefinition{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: csiapiv1alpha1.CsiDriverResourcePlural + "." + csiapiv1alpha1.GroupName,
+		},
+		Spec: apiextensionsv1beta1.CustomResourceDefinitionSpec{
+			Group:   csiapiv1alpha1.GroupName,
+			Version: csiapiv1alpha1.SchemeGroupVersion.Version,
+			Scope:   apiextensionsv1beta1.ClusterScoped,
+			Names: apiextensionsv1beta1.CustomResourceDefinitionNames{
+				Plural: csiapiv1alpha1.CsiDriverResourcePlural,
+				Kind:   reflect.TypeOf(csiapiv1alpha1.CSIDriver{}).Name(),
+			},
+		},
+	}
+	res, err := adc.crdClient.ApiextensionsV1beta1().CustomResourceDefinitions().Create(crd)
+
+	if err != nil && !apierrors.IsAlreadyExists(err) {
+		glog.Errorf("failed to create CSIDrivers CRD: %#v, err: %#v",
+			res, err)
+	} else if apierrors.IsAlreadyExists(err) {
+		glog.Warningf("CSIDrivers CRD already exists: %#v, err: %#v",
+			res, err)
+	} else {
+		glog.Infof("CSIDrivers CRD created successfully: %#v",
+			res)
+	}
+
+	crd = &apiextensionsv1beta1.CustomResourceDefinition{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: csiapiv1alpha1.CsiNodeInfoResourcePlural + "." + csiapiv1alpha1.GroupName,
+		},
+		Spec: apiextensionsv1beta1.CustomResourceDefinitionSpec{
+			Group:   csiapiv1alpha1.GroupName,
+			Version: csiapiv1alpha1.SchemeGroupVersion.Version,
+			Scope:   apiextensionsv1beta1.ClusterScoped,
+			Names: apiextensionsv1beta1.CustomResourceDefinitionNames{
+				Plural: csiapiv1alpha1.CsiNodeInfoResourcePlural,
+				Kind:   reflect.TypeOf(csiapiv1alpha1.CSINodeInfo{}).Name(),
+			},
+		},
+	}
+	res, err = adc.crdClient.ApiextensionsV1beta1().CustomResourceDefinitions().Create(crd)
+
+	if err != nil && !apierrors.IsAlreadyExists(err) {
+		glog.Errorf("failed to create CSINodeInfo CRD: %#v, err: %#v",
+			res, err)
+	} else if apierrors.IsAlreadyExists(err) {
+		glog.Warningf("CSINodeInfo CRD already exists: %#v, err: %#v",
+			res, err)
+	} else {
+		glog.Infof("CSINodeInfo CRD created successfully: %#v",
+			res)
+	}
+
+	return nil
```

Comment:
How well will the rest of the system cope if only one of these is installed (say the other one has a transient error)?

Reply:
Different parts of the volume plugin provisioning/attach/mount process look at CRs for one of these CRDs. If one of the CRDs fails to create, it would just change where in the provision/attach/mount process subsequent failure will happen.

Since it is better to fail fast, I changed the errors here to return error instead of silently log.